README
NAME
    Gentoo::MetaEbuild::Spec::Base - A Base Class for Gentoo MetaEbuild
    Specifications.

VERSION
    version 1.000002

SYNOPSIS
        use Gentoo::MetaEbuild::Spec::Base; # or some derived class
        Gentoo::MetaEbuild::Spec::Base->check( $datastructure );

    This base-class only validates the most basic of basic, that the data is
    a { } using Data::Rx and using the shipped File::ShareDir v1.0.0.json
    spec to do that.

    This will be more practical in consuming classes as they'll override
    selected methods/ship different spec files, but maintain the same useful
    interface.

METHODS
  check
        Packagename->check( $datastructure );

        Packagename->check( $datastructure, \%opts );

        Packagename->check( $datastructure, { version => '0.1.0' });

EXTENDING
    Extending should be this simple:

        package FooBarBaz;
        use Moose;
        extends 'Gentoo::MetaEbuild::Spec::Base';

        1;

    and then ship a directory of Data::Rx spec files as the Module ShareDir
    for that module.

TESTING
    The only fun thing with testing is the File::ShareDir directory hasn't
    been installed yet, but its simple to get around.

        use FindBin;
        use Path::Tiny qw( path );
        use Gentoo::MetaEbuild::Spec::Base;

        Gentoo::MetaEbuild::Spec::Base->_spec_dir(
            path($FindBin::Bin)->parent->child('share')
        );

        # Code as per usual.

        my $shareroot = path($FindBin::Bin)->parent();

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric <kentnl@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Gentoo-MetaEbuild-Spec-Base

1.000003 2017-03-10T10:11:59Z ad63d57

1.000002 2017-03-10T10:10:34Z becb985
 - Bugfix: Avoid failures with -Ddefault_inc_excludes_dot
 - Support for older Test::More

 [Dependencies::Stats]
 - Dependencies changed since 1.000001, see misc/*.deps* for details
 - configure: +1 (recommends: +1 ↑1)
 - develop: +15 ↑4 -3 (recommends: +1, suggests: ↑2)
 - runtime: ↓1 -1 (recommends: +2)
 - test: +1 ↓1 (recommends: ↑1 ↓1)

1.000001 2014-08-15T23:02:42Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - whitespace reduced.

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - develop: +1 (suggests: ↑1)

 [Misc]
 - Improve source side POD to reduce build side whitespace.

1.000000 2014-08-10T11:08:49Z
 [00 Major]
 - Significant changes that may affect hand-operated downstreams.

 [01 Version Scheme]
 - Migrated from x.y.z to x.yyyzzz
 - This is the primary reason for the 1.x on the box
 - https://questhub.io/realm/perl/quest/5388d3a9bbd0be2224000129

 [02 Tooling Change]
 - Moved to ExtUtils::MakeMaker
 - Tookit dependencies for "latest" things now only "recommended"
 - https://questhub.io/realm/perl/quest/5388cc12bbd0be5b2500011f

 [03 Some critical components changed]
 - Path::Class -> Path::Tiny
 - JSON::XS -> JSON::MaybeXS
 - Test::File::ShareDir -> Test::File::ShareDir::Module`

 [Dependencies::Stats]
 - Dependencies changed since 0.1.4, see misc/*.deps* for details
 - build: -1
 - configure: +2 -1 (recommends: +1)
 - develop: +48 -1 (recommends: -1, suggests: +1 ↑1)
 - runtime: +4 ↓1 -3
 - test: +3 -3 (recommends: +3)

 [Misc]
 - Report versions moved to Test::ReportPrereqs
 - Compile test moved to ::PerFile
 - Repo moved to kentnl org

0.1.4 2012-01-04T05:16:00Z
 - Maintenance release.

 [Dependencies::Stats]
 - Dependencies changed since 0.1.3, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - runtime: +3
 - test: -1

 [Internals]
 - All namespaces declare AUTHORITY
 - VERSION no longer declared in BEGIN

 [Packaging]
 - migrate Changes format to CPAN::Changes
 - Update Copyright years, formatting and address.
 - Move author/release tests to xt/
 - Repository urls moved to https://
 - x_authority added to metadata.

0.1.3 2011-02-21T13:40:12Z
 - Move tests to use Test::File::ShareDir

 [Dependencies::Stats]
 - Dependencies changed since 0.1.2, see misc/*.deps* for details
 - test: +1

0.1.2 2011-02-21T09:38:04Z
 - Fix ->check( .., { version => }) not working at all.
 - Add tests for this behaviour.
 - Ensure to test what happens when a given spec is non-existant.
 - Add support for { version => } being a VersionObject.

 [Dependencies::Stats]
 - Dependencies changed since 0.1.1, see misc/*.deps* for details
 - runtime: +1
 - test: +1

0.1.1 2011-02-21T07:52:03Z
 - Fix pesky module_dir being missing. Need a better way to test this -_-

0.1.0 2011-02-21T07:12:09Z
 - First version, released on an unsuspecting world.


lib/Gentoo/MetaEbuild/Spec/Base.pm
use 5.006;    # our
use strict;
use warnings;

package Gentoo::MetaEbuild::Spec::Base;

our $VERSION = '1.000003';

# ABSTRACT: A Base Class for Gentoo MetaEbuild Specifications.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Moose;
use MooseX::ClassAttribute qw( class_has );

use File::ShareDir qw( module_dir );
use Path::Tiny qw( path );
use MooseX::Types::Moose qw( Str CodeRef );
use MooseX::Types::Perl qw( VersionObject );
use MooseX::Types::Path::Tiny qw( AbsPath AbsDir );
use Scalar::Util qw( blessed );
use MooseX::Has::Sugar qw( ro lazy_build rw coerce lazy );
use version;

use namespace::autoclean;

class_has '_decoder' => (
  isa => CodeRef,
  ro, lazy_build,
  traits  => [qw( Code )],
  handles => { _decode => 'execute', },
);

sub _build__decoder {
  require JSON::MaybeXS;
  my $decoder = JSON::MaybeXS->new()->utf8(1)->relaxed(1);
  return sub {
    $decoder->decode(shift);
  };
}

class_has '_spec_dir' => (
  isa => AbsDir,
  rw, lazy_build,
);

sub _build__spec_dir {
  my ($self) = shift;
  my $classname;
  if ( ref $self && blessed $self ) {
    $classname = blessed $self;
  }
  elsif ( ref $self ) {
    require Carp;
    ## no critic (ValuesAndExpressions::RequireInterpolationOfMetachars)
    Carp::confess q{$_[0] is not a Class/Object};
  }
  else {
    $classname = $self;
  }
  return path( module_dir($classname) );
}

class_has '_version' => (
  isa => VersionObject,
  coerce, ro, lazy, default => sub { q{0.1.0} },
);

class_has '_extension' => (
  isa => Str,
  ro, lazy, default => sub { q{.json} },
);

class_has '_schema_creator' => (
  isa => CodeRef,
  ro, lazy_build,
  traits  => [qw( Code )],
  handles => { _make_schema => 'execute', },
);

__PACKAGE__->meta->make_immutable;
no Moose;

sub _build__schema_creator {
  require Data::Rx;
  my $rx = Data::Rx->new();
  return sub {
    $rx->make_schema(shift);
  };
}

sub _opt_check {
  my ( $self, $opts ) = @_;
  if ( not exists $opts->{version} ) {
    $opts->{version} = $self->_version;
  }
  elsif ( blessed $opts->{version} ) {

  }
  else {
    $opts->{version} = version->parse( $opts->{version} );
  }
  return $opts;
}

sub _spec_file {
  my ( $self, $opts ) = @_;
  $opts = $self->_opt_check($opts);
  return $self->_spec_dir->child( $opts->{version}->normal . $self->_extension );
}

sub _spec_data {
  my ( $self, $opts ) = @_;
  $opts = $self->_opt_check($opts);
  return $self->_decode( scalar $self->_spec_file($opts)->slurp() );
}

sub _schema {
  my ( $self, $opts ) = @_;
  $opts = $self->_opt_check($opts);
  return $self->_make_schema( $self->_spec_data($opts) );
}











sub check {
  my ( $self, $json_data, $opts ) = @_;
  $opts = $self->_opt_check($opts);
  return $self->_schema($opts)->check($json_data);
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Gentoo::MetaEbuild::Spec::Base - A Base Class for Gentoo MetaEbuild Specifications.

=head1 VERSION

version 1.000003

=head1 SYNOPSIS

    use Gentoo::MetaEbuild::Spec::Base; # or some derived class
    Gentoo::MetaEbuild::Spec::Base->check( $datastructure );

This base-class only validates the most basic of basic, that the data is a  { } using Data::Rx
and using the shipped File::ShareDir v1.0.0.json spec to do that.

This will be more practical in consuming classes as they'll override selected methods/ship different spec files,
but maintain the same useful interface.

=head1 METHODS

=head2 check

    Packagename->check( $datastructure );

    Packagename->check( $datastructure, \%opts );

    Packagename->check( $datastructure, { version => '0.1.0' });

=head1 EXTENDING

Extending should be this simple:

    package FooBarBaz;
    use Moose;
    extends 'Gentoo::MetaEbuild::Spec::Base';

    1;

and then ship a directory of Data::Rx spec files as the Module ShareDir for that module.

=head1 TESTING

The only fun thing with testing is the File::ShareDir directory hasn't been installed yet, but its simple to get around.

    use FindBin;
    use Path::Tiny qw( path );
    use Gentoo::MetaEbuild::Spec::Base;

    Gentoo::MetaEbuild::Spec::Base->_spec_dir(
        path($FindBin::Bin)->parent->child('share')
    );

    # Code as per usual.

    my $shareroot = path($FindBin::Bin)->parent();

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4206"
  },
  "configure_requires" => {
    "Module::Build" => "0.4206"
  },
  "dist_abstract" => "A Base Class for Gentoo MetaEbuild Specifications.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Gentoo-MetaEbuild-Spec-Base",
  "dist_version" => "0.1.5",
  "license" => "perl",
  "module_name" => "Gentoo::MetaEbuild::Spec::Base",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Data::Rx" => 0,
    "File::ShareDir" => 0,
    "JSON::XS" => 0,
    "Moose" => 0,
    "MooseX::ClassAttribute" => 0,
    "MooseX::Has::Sugar" => 0,
    "MooseX::Types::Moose" => 0,
    "MooseX::Types::Path::Class" => 0,
    "MooseX::Types::Perl" => 0,
    "Path::Class" => 0,
    "Scalar::Util" => 0,
    "namespace::autoclean" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "version" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "share_dir" => {
    "module" => {
      "Gentoo::MetaEbuild::Spec::Base" => "share"
    }
  },
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Test::Fatal" => 0,
    "Test::File::ShareDir" => 0,
    "Test::More" => "1.001003"
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "Module::Build" => "0.4206",
  "Test::Fatal" => 0,
  "Test::File::ShareDir" => 0,
  "Test::More" => "1.001003"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

use File::ShareDir::Install;
$File::ShareDir::Install::INCLUDE_DOTFILES = 1;
$File::ShareDir::Install::INCLUDE_DOTDIRS = 1;
install_share module => "Gentoo::MetaEbuild::Spec::Base", "share";


my %WriteMakefileArgs = (
  "ABSTRACT" => "A Base Class for Gentoo MetaEbuild Specifications.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::ShareDir::Install" => "0.06"
  },
  "DISTNAME" => "Gentoo-MetaEbuild-Spec-Base",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Gentoo::MetaEbuild::Spec::Base",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Rx" => 0,
    "File::ShareDir" => 0,
    "JSON::MaybeXS" => 0,
    "Moose" => 0,
    "MooseX::ClassAttribute" => 0,
    "MooseX::Has::Sugar" => 0,
    "MooseX::Types::Moose" => 0,
    "MooseX::Types::Path::Tiny" => 0,
    "MooseX::Types::Perl" => 0,
    "Path::Tiny" => 0,
    "Scalar::Util" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "version" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Test::Fatal" => 0,
    "Test::File::ShareDir::Module" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.000003",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Data::Rx" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::ShareDir" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "JSON::MaybeXS" => 0,
  "Moose" => 0,
  "MooseX::ClassAttribute" => 0,
  "MooseX::Has::Sugar" => 0,
  "MooseX::Types::Moose" => 0,
  "MooseX::Types::Path::Tiny" => 0,
  "MooseX::Types::Perl" => 0,
  "Path::Tiny" => 0,
  "Scalar::Util" => 0,
  "Test::Fatal" => 0,
  "Test::File::ShareDir::Module" => 0,
  "Test::More" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}
